### PR TITLE
Noms 54 test js

### PR DESCRIPTION
**Pull request by corydodt**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -9,5 +9,6 @@ before_install:
     - pip install -U pip setuptools
 install:
     - pip install -r requirements.txt
+    - npm install
```
Comment (katharinechen) [inline .travis.yml]: Hm, I am a little bit confused. So `npm install` without any arguments Install the dependencies in the local `node_modules` folder. 

How do I see the contents of my local `node_modules` folder? Is travis looking at my local `node_modules` folder or does it have it's own? If it is different, how do I see the folder travis uses? 

Lastly, is it important to use the `--production` flag when we do this install?

File: docs/devgettingstarted.rst [hunk 2]
```diff
@@ -129,6 +130,25 @@ You can ALSO restart noms by running::
 
     docker kill -s HUP deployment_noms-main_1
 
+Running Tests
```
Comment (katharinechen) [inline docs/devgettingstarted.rst]: Do we want to talk about `pyflakes` and `jshint` here?

File: bin/do-travis-build [hunk 3]
```diff
@@ -12,6 +12,8 @@ eval "$(whisk describe)"
 set +a
 
 # run tests
+jshint static && echo "jshint ok 👌"
```
Comment (katharinechen) [inline bin/do-travis-build]: I believe at work, we swap `jslint` for `eslint`? Does the rationale for that decision apply here?

File: .travis.yml (hunk 1 above)
Comment (corydodt) [reply .travis.yml]: `npm install` installs the packages in packages.json. It *creates* node_modules. (And updates it.)

`npm ls` will list your node_modules, but it's just a directory of files, you can look around yourself. Travis creates everything from scratch every time, it has no idea what's on your computer. Again, node_modules doesn't exist until npm install creates it from packages.json.

`--production` installs only packages in the main section of packages.json and ignores any in devDependencies. We don't care about this at all.

File: docs/devgettingstarted.rst (hunk 2 above)
Comment (corydodt) [reply docs/devgettingstarted.rst]: I will probably replace jshint with eslint (I could probably do that in the course of this case, it wouldn't take long at all).

We should mention them here.

File: bin/do-travis-build (hunk 3 above)
Comment (corydodt) [reply bin/do-travis-build]: The main reason for the switch was we were using es7 features at work. We aren't doing that for noms yet, but I think it's still a good idea to switch.

File: docs/devgettingstarted.rst (hunk 2 above)
Comment (katharinechen) [reply docs/devgettingstarted.rst]: Okay, sounds good.

File: package.json [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  "dependencies": {
+    "angular": "^1.6.8",
+    "angular-mocks": "^1.6.7",
+    "chai": "^4.1.2",
+    "html2js": "^0.2.0",
+    "karma": "^1.7.1",
+    "karma-chai": "^0.1.0",
+    "karma-chrome-launcher": "^2.2.0",
+    "karma-coverage": "^1.1.1",
+    "karma-html2js-preprocessor": "^1.1.0",
```
Comment (katharinechen) [inline package.json]: When I `npm install`, I get an error for this particular install. 

`npm ERR! invalid: karma-html2js-preprocessor@1.1.0 /Users/katharinechen/Noms/node_modules/karma-html2js-preprocessor`

File: .travis.yml (hunk 1 above)
Comment (katharinechen) [reply .travis.yml]: Oh, I misunderstood the documentation. Okay, that makes sense.

File: bin/do-travis-build [hunk 5]
```diff
@@ -3,7 +3,7 @@
 set -euo pipefail
 IFS=$'\n\t'
 
-PATH="$PATH:./bin"
+PATH="$PATH:./bin:./node_modules/.bin"
```
Comment (katharinechen) [inline bin/do-travis-build]: Why does the `PATH` need to change?

File: bin/do-travis-build [hunk 6]
```diff
@@ -12,6 +12,8 @@ eval "$(whisk describe)"
 set +a
 
 # run tests
+jshint static && echo "jshint ok 👌"
+karma start
```
Comment (katharinechen) [inline bin/do-travis-build]: I had run `npm install -g karma-cli` before I can actually run `karma start`. Is that something that we need to add to `package.json`?

File: static/js/app.js [hunk 7]
```diff
@@ -1,4 +1,3 @@
-// jshint -W097
```
Comment (katharinechen) [inline static/js/app.js]: Ah! We are so close to having 100% branch coverage as well! Just something in this file!

File: static/js/controllers/recipenew.js [hunk 8]
```diff
@@ -1,25 +1,31 @@
...
-      // temporary
-      recipe = Object.assign({}, recipe);
-      recipe.ingredients = [recipe.ingredients];
-      recipe.instructions = [recipe.instructions];
+        // temporary
+        recipe = Object.assign({}, recipe);
+        recipe.ingredients = [recipe.ingredients];
+        recipe.instructions = [recipe.instructions];
+
+        $http.post('/api/recipe/create', recipe).then(
```
Comment (katharinechen) [inline static/js/controllers/recipenew.js]: Why the change?

File: package.json (hunk 4 above)
Comment (corydodt) [reply package.json]: ok, investigating

File: bin/do-travis-build (hunk 5 above)
Comment (corydodt) [reply bin/do-travis-build]: No, it's because of the programs installed by `npm install` :)

We use several tools written in ES, and those tools get on your machine by installing them with `npm install`. The command-line executable that goes with each one is put in the directory printed by running `npm bin`, which is almost always `node_modules/.bin`. So you're just making those commands available in your shell.

File: docs/devgettingstarted.rst [hunk 9]
```diff
@@ -129,6 +130,25 @@ You can ALSO restart noms by running::
...
+
+There are a few different ways to use pytest:
+
+- To run a specific test, use: ``pytest noms/test/test_rendering.py``
+- To run all of the test, use: ``pytest``
+- To run only the failing test, use: ``pytest --lf``
+
+The foreground test runner is ``karma``. To run these tests, run::
+
+.. code-block:: bash
```
[the remark this answers is not in the document]
Comment (corydodt) [reply docs/devgettingstarted.rst]: Pretty sure I was copying syntax from you? IHNI how to test on RtD before we land a case.

File: bin/do-travis-build (hunk 6 above)
Comment (corydodt) [reply bin/do-travis-build]: I don't know, I'll start from scratch and see what happens.

File: static/js/app.js (hunk 7 above)
Comment (corydodt) [reply static/js/app.js]: If we're going to do branch coverage, we should have a case where we do it for both python and js.

File: static/js/controllers/recipenew.js (hunk 8 above)
Comment (corydodt) [reply static/js/controllers/recipenew.js]: Node tells me that `Promise.success()` is deprecated, and we should use `Promise.then()`, which has this syntax.

File: docs/devgettingstarted.rst (hunk 2 above)
Comment (corydodt) [reply docs/devgettingstarted.rst]: Switch to eslint is done.

Comment (corydodt) [reply docs/devgettingstarted.rst]: Updated docs

File: package.json (hunk 4 above)
Comment (corydodt) [reply package.json]: I didn't have this problem.

Can you show me the output of `node --version` and `npm --version`? Here's mine:

```
corys-MacBook-Pro:Noms cory$ node --version
v8.9.1
(noms) ----------------------------------------------------------
corys-MacBook-Pro:Noms cory$ npm --version
5.6.0
```

File: docs/devgettingstarted.rst [hunk 10]
```diff
@@ -129,6 +130,42 @@ You can ALSO restart noms by running::
...
+- To run only the failing test, use: ``pytest --lf``
+
+To run ``pyflakes``, just run::
+
+    pyflakes
+
+*For ECMAScript*
+
+- The foreground test runner is ``karma``. To run these tests, run::
+- We also run ``eslint`` to catch common errors.
```
Comment (katharinechen) [inline docs/devgettingstarted.rst]: `/Users/katharinechen/Noms/docs/devgettingstarted.rst:160: WARNING: Literal block expected; none found.`

Comment (katharinechen) [reply docs/devgettingstarted.rst]: Hm, there seems to be another warning that I haven't seen before. `copying static files... WARNING: html_static_path entry u'/Users/katharinechen/Noms/docs/_static' does not exist.`

File: docs/dod.rst [hunk 11]
```diff
@@ -1,30 +1,19 @@
...
-Definition of Done 
+Definition of Done
 ------------------
-Definition of Done for a noms case: 
+Definition of Done for a noms case:
 
 - There should be 100% test coverage for both python and javascript.
-- All automated test pass.
+- All automated tests pass. See :doc:`devgettingstarted` for details on writing and running tests.
+-
```
Comment (katharinechen) [inline docs/dod.rst]: extra bullet

File: docs/devgettingstarted.rst [hunk 12]
```diff
@@ -129,6 +130,42 @@ You can ALSO restart noms by running::
...
 
+Running Tests
+~~~~~~~~~~~~~
+Noms uses several test runners. All tools listed here are run by travis during
+the build, and must pass 100% for the build to succeed, including code coverage
+where appropriate.
+
+*For Python:*
+- ``pytest`` as its backend test runner, and all tests are written in the ``pytest`` style.
+- Python also uses ``pyflakes`` to catch common errors.
```
Comment (katharinechen) [inline docs/devgettingstarted.rst]: Weird. I don't know why when I run `pyflakes`, it just stalls.

File: noms/test/conftest.py [hunk 13]
```diff
@@ -119,7 +119,7 @@ def mockConfig(mockDatabase):
 
     try:
         cols = mockDatabase.collection_names()
-        for c in cols:
+        for c in cols: # pragma: nocover
```
Comment (katharinechen) [inline noms/test/conftest.py]: If there is no cover for a test? Does that mean it is not being used? If so why do we need it?

File: package.json (hunk 4 above)
Comment (katharinechen) [reply package.json]: Hm, weird. I stop getting this issue.

File: .eslintrc.json [hunk 14]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (katharinechen) [inline .eslintrc.json]: I don't understand why we need three different `eslintrc.json` files?

File: docs/devgettingstarted.rst (hunk 10 above)
Comment (corydodt) [reply docs/devgettingstarted.rst]: Need to chec\k this out

File: docs/devgettingstarted.rst (hunk 12 above)
Comment (corydodt) [reply docs/devgettingstarted.rst]: You're supposed to run pyflakes with 1 or more directory names, e.g. `pyflakes noms`

--BUT--pytest runs pyflakes for you with the `--flakes` option which is automatically included from `pytest.ini`.

This appears to be working correctly.

Comment (corydodt) [reply docs/devgettingstarted.rst]: Just noticed that the documentation is wrong, so I'm fixing it.

File: noms/templates/universal-header.html [hunk 15]
```diff
@@ -13,7 +13,7 @@
   
 <!-- data preloaded by the server, becomes part of the Noms scope in angular -->
 <script>
-<% set nomsPreload = preload or none %>
+<!-- <% set nomsPreload = preload or none %> -->
```
[the remark this answers is not in the document]
Comment (corydodt) [reply noms/templates/universal-header.html]: OK, to explain this one: 

- `<% %>` is a jinja directive, so it's interpreted on the server before the page is even sent to the browser
- `<!-- --!>` is an HTML comment, meaning it's only interpreted on the browser, not on the server
- Therefore, the directive on this line runs on the server regardless of the comment markup around it
- in running, it has a chance of producing some output, which appears in your markup, and can cause problems. You don't want or care about that output, you really only want the side effect of the `<% set %>` directive (making a local variable available to the rest of the template)
- Therefore, putting HTML comment syntax around it prevents the undesirable output from causing any problems in the browser, _without_ interfering with the directive's behavior on the server.

File: noms/test/conftest.py (hunk 13 above)
Comment (corydodt) [reply noms/test/conftest.py]: `nocover` just prevents it from being counted in the line total that coverage uses to determine percentage. It doesn't mean the code doesn't run.

It's been a while since I added this, but iirc what was happening was that the database would sometimes be empty after running tests, and therefore there was no need to do this cleanup step. `conftest.py` is only used in tests anyway, so ignoring this line of coverage doesn't bias application test coverage in any way.

File: .eslintrc.json (hunk 14 above)
Comment (corydodt) [reply .eslintrc.json]: Each one has a scope which is limited to the directory (and subdirectories) it's a part of.

Having more than one of them allows us to fine tune our desired eslint behavior to the part of the application it's associated with. Decoupling them by only having one gives us less control. (For example, we might want warnings for X in one directory but not in another.)

File: docs/devgettingstarted.rst (hunk 10 above)
Comment (corydodt) [reply docs/devgettingstarted.rst]: This needed a minor change to conf.py is all.
